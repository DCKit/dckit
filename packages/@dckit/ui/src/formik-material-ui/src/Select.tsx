import * as React from 'react';
import MuiSelect, {
  SelectProps as MuiSelectProps,
} from '@material-ui/core/Select';
import {
  FieldInputProps,
  FieldMetaProps,
  FieldHelperProps,
  useField,
  useFormikContext,
} from 'formik';

export interface SelectProps extends Omit<MuiSelectProps, 'name' | 'value'> {
  name: string;
}

export function useFieldToSelect<Val = unknown>(
  { disabled, name, ...props }: SelectProps,
  customize?: (
    props: [FieldInputProps<Val>, FieldMetaProps<Val>, FieldHelperProps<Val>]
  ) => Partial<Omit<SelectProps, 'name'>>
): MuiSelectProps {
  const { isSubmitting } = useFormikContext();
  const fieldProps = useField(name);

  const [field, , helpers] = fieldProps;

  const onChange = React.useCallback(
    (event: React.ChangeEvent<{ value: unknown }>) => {
      // Special case for multiple and native
      if (props.multiple && props.native) {
        const { options } = event.target as HTMLSelectElement;
        const value: string[] = [];
        for (let i = 0, l = options.length; i < l; i += 1) {
          if (options[i].selected) {
            value.push(options[i].value);
          }
        }

        helpers.setValue(value);
      } else {
        field.onChange(event);
      }
    },
    [field.name, props.multiple, props.native]
  );

  return {
    disabled: disabled ?? isSubmitting,
    ...props,
    ...field,
    onChange,
    ...customize?.(fieldProps),
  };
}

export function Select(props: SelectProps) {
  return <MuiSelect {...useFieldToSelect(props)} />;
}

Select.displayName = 'FormikMaterialUISelect';
