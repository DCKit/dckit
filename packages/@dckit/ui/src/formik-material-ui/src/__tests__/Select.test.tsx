import * as React from 'react';
import { Formik, Form } from 'formik';
import renderer from 'react-test-renderer';
import MenuItem from '@material-ui/core/MenuItem';

import { Select } from '../Select';

test('Select Renders Correctly', () => {
  const component = renderer.create(
    <Formik onSubmit={() => {}} initialValues={{ test: '' }}>
      <Form>
        <Select name="test">
          <MenuItem value="">
            <em>None</em>
          </MenuItem>
          <MenuItem value={10}>Ten</MenuItem>
          <MenuItem value={20}>Twenty</MenuItem>
          <MenuItem value={30}>Thirty</MenuItem>
        </Select>
      </Form>
    </Formik>
  );

  expect(component.toJSON()).toMatchSnapshot();
});
