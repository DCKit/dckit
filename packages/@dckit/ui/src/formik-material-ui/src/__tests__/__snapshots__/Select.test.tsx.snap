// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Renders Correctly 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline"
    onClick={[Function]}
  >
    <div
      aria-haspopup="listbox"
      aria-labelledby=" mui-component-select-test"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      id="mui-component-select-test"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#8203;",
          }
        }
      />
    </div>
    <input
      name="test"
      onAnimationStart={[Function]}
      type="hidden"
      value=""
    />
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root MuiSelect-icon"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
  </div>
</form>
`;
